rollbar_test.go

```diff
@@ -402,58 +405,120 @@ func (cs cs) Stack() []runtime.Frame {
+		if ok {
+			t.Errorf("standard errors should not be handled")
+		}
+	})
+	t.Run("Stacker", func(t *testing.T) {
+		trace := getCallersFrames(0)
+		err := cs{fmt.Errorf("cause"), nil, trace}
+		extractedTrace, ok := DefaultStackTracer(err)
+		if extractedTrace == nil {
+			t.Error("Stackers should return a trace")
+		} else if extractedTrace[0] != trace[0] {
+			t.Error("the trace from the error must be extracted")
+		}
+		if !ok {
+			t.Error("Stackers should be handled")
+		}
+	})
 }
 
-func TestGetOrBuildStackOfCauseStackerWithoutParent(t *testing.T) {
-	cause := fmt.Errorf("cause")
-	effect := cs{fmt.Errorf("effect"), cause, getCallersFrames(0)}
-	if len(effect.Stack()) == 0 {
-		t.Fatal("stack should not be empty")
-	}
-	if effect.Stack()[0] != getOrBuildFrames(effect, nil, 0, nil)[0] {
-		t.Error("should use stack from effect")
-	}
-}
+func TestGetOrBuildFrames(t *testing.T) {
+	// These tests all use the default stack tracer. The logic this is testing doesn't really
+	// depend on how the stack trace is extracted.
 
-func TestGetOrBuildStackOfCauseStackerWithParent(t *testing.T) {
-	cause := fmt.Errorf("cause")
-	effect := cs{fmt.Errorf("effect"), cause, getCallersFrames(0)}
-	effect2 := cs{fmt.Errorf("effect2"), effect, getCallersFrames(0)}
-	if effect2.Stack()[0] != getOrBuildFrames(effect2, effect, 0, nil)[0] {
-		t.Error("should use stack from effect2")
-	}
+	t.Run("standard error without parent", func(t *testing.T) {
+		err := fmt.Errorf("")
+		trace := getOrBuildFrames(err, nil, 0, DefaultStackTracer)
+		if nil == trace {
+			t.Error("should build a new stack trace if error has no stack and parent is nil")
+		}
+	})
+	t.Run("standard error with traceable parent", func(t *testing.T) {
+		cause := fmt.Errorf("cause")
+		effect := cs{fmt.Errorf("effect"), cause, getCallersFrames(0)}
+		if nil != getOrBuildFrames(cause, effect, 0, DefaultStackTracer) {
+			t.Error("should return nil if child is not traceable but parent is")
+		}
+	})
+	t.Run("standard error with non-traceable parent", func(t *testing.T) {
+		child := fmt.Errorf("child")
+		parent := uw{fmt.Errorf("parent"), child}
+		trace := getOrBuildFrames(child, parent, 0, DefaultStackTracer)
+		if nil == trace {
+			t.Error("should build a new stack trace if parent is not traceable")
+		}
+	})
+	t.Run("traceable error without parent", func(t *testing.T) {
+		cause := fmt.Errorf("cause")
+		effect := cs{fmt.Errorf("effect"), cause, getCallersFrames(0)}
+		if effect.Stack()[0] != getOrBuildFrames(effect, nil, 0, DefaultStackTracer)[0] {
+			t.Error("should use stack trace from effect")
+		}
+	})
+	t.Run("traceable error with traceable parent", func(t *testing.T) {
+		cause := fmt.Errorf("cause")
+		effect := cs{fmt.Errorf("effect"), cause, getCallersFrames(0)}
+		effect2 := cs{fmt.Errorf("effect2"), effect, getCallersFrames(0)}
+		if effect.Stack()[0] != getOrBuildFrames(effect, effect2, 0, DefaultStackTracer)[0] {
+			t.Error("should use stack from child, not parent")
+		}
+	})
```

Comment:
This test is equivalent to the old `TestGetOrBuildStackOfCauseStackerWithParent`, with one important change. In the original, the code was as follows:

```go
cause := fmt.Errorf("cause")
effect := cs{fmt.Errorf("effect"), cause, getCallersFrames(0)}
effect2 := cs{fmt.Errorf("effect2"), effect, getCallersFrames(0)}
if effect2.Stack()[0] != getOrBuildFrames(effect2, effect, 0, nil)[0] {
	t.Error("should use stack from effect2")
}
```

You can see that `effect2` wraps `effect`, which wraps `cause`. So `effect2` is the parent of `effect`. But, `getOrBuildFrames` passed `effect2` as the child and `effect` as the parent.

Since I believe this to be an incorrect test, I have swapped `effect2` and `effect` and modified the test to assert that it yields the stack from `effect`.

Reply:
It's worth noting that I don't think this actually changes anything about the logic tested by this case. Both `effect` and `effect2` are traceable, which means `getOrBuildFrames` would just return the stack from whichever one was passed as the child.